lib/classes/PluginManager.js

```diff
@@ -309,6 +310,27 @@ class PluginManager {
     });
   }
 
+  loadVariableGetters(pluginInstance) {
+    for (const [variablePrefix, variableGetter] of _.entries(
+      pluginInstance.variableGetters || {}
+    )) {
+      this.serverless.variables.customVariableResolverFuncs[variablePrefix] = variableGetter.bind(
+        this.serverless.variables
+      );
+      this.serverless.variables.variableResolvers.push([
+        new RegExp(`^${variablePrefix}:`),
+        variablePrefix,
```

Comment:
This approach is prone to collisions, as if e.g. we introduce a `tracker` variable service.. then it'll collide with [`variables.tracker`](https://github.com/serverless/serverless/blob/55a794f4e07cd7a903ee80c4b68f9f4818b40355/lib/classes/Variables.js#L44) property, and crash here: https://github.com/serverless/serverless/blob/55a794f4e07cd7a903ee80c4b68f9f4818b40355/lib/classes/Variables.js#L557-L558

To avoid such issues, and improve readability of that handling I would rather populate `this.serverless.variables.variableResolvers` with hashes where resolver is referenced directly as e.g.:

```javascript
{
  matchingRegExp: new RegExp(`^${variablePrefix}:`),
  resolve: variableGetter.bind(this.serverless.variables)
}
```

Reply:
So, the reason for going with names is because of how the disabling during prepopulation works. Whatever solution we do has to support monkey patching: https://github.com/serverless/serverless/pull/6566/files#diff-b921cbcedd2151d24391fb39acb84adeR108-R114

Reply:
What about just adding a check for conflicts when registering variable getters? eg, it can't be in the existing list of resolvers, or it throws an error.

Reply:
>  Whatever solution we do has to support monkey patching: https://github.com/serverless/serverless/pull/6566/files#diff-b921cbcedd2151d24391fb39acb84adeR108-R114

In case of `this[dependentService.method]` it's different as we use function names only for _fixed_ and  _known_ set of methods that is decided internally.
And custom resolvers (as added by users) are moved to other safe namespace (`this.customVariableResolverFuncs`), so there's no risk of collision here.

Here I just question using similar method for variable resolvers, as it seems to have issues. Why going the route I proposed feels problematic? I think it's a way safer pattern and also way more consumable (when code readability is concerned)

Reply:
I like the idea of storing the function itself, I just don't know how to do the equivalent monkey patching to disable dependent services when doing that. I'm open to suggestions ;)

Reply:
I take the problem is with this part of `disableDepedentServices`: https://github.com/serverless/serverless/blob/55a794f4e07cd7a903ee80c4b68f9f4818b40355/lib/classes/Variables.js#L120
where we rely on fact that methods are referenced as strings in `variableResolvers`

I think we can fill `dependentServices` array with  same  objects (same instances) that go into `variableResolvers`. Then in `disableDepedentServices` you may override `resolver` in those objects with _noop_ for a meantime.
It'll require change within existing logic, but I guess it's for good.